client/modules/IDE/components/Editor/index.jsx

```diff
@@ -141,546 +228,443 @@ class Editor extends React.Component {
+          Right: (cmm, e) => {
+            setFocusedLinkElement(true);
+          },
+          Left: (cmm, e) => {
+            removeFocusedLinkElement();
+          },
+          Up: (cmm, e) => {
+            const onLink = removeFocusedLinkElement();
+            e.moveFocus(-1);
+            setFocusedLinkElement(onLink);
+          },
+          Down: (cmm, e) => {
+            const onLink = removeFocusedLinkElement();
+            e.moveFocus(1);
+            setFocusedLinkElement(onLink);
+          },
+          Enter: (cmm, e) => {
+            if (focusedLinkElement) focusedLinkElement.click();
+            else e.pick();
+          }
+        },
+        closeOnUnfocus: false
+      };
+      if (cm.current.options.mode === 'javascript') {
+        // JavaScript
+        CodeMirror.showHint(
+          cm.current,
+          () => {
+            const c = cm.current.getCursor();
+            const token = cm.current.getTokenAt(c);
+            const hints = hinterar
+              .search(token.string)
+              .filter((h) => h.item.text[0] === token.string[0]);
+            return {
+              list: hints,
+              from: CodeMirror.Pos(c.line, token.start),
+              to: CodeMirror.Pos(c.line, c.ch)
+            };
+          },
+          hintOptions
+        );
+      } else if (cm.current.options.mode === 'css') {
+        // CSS
+        CodeMirror.showHint(cm.current, CodeMirror.hint.css, hintOptions);
+      }
+    };
 
-    this._cm.on('keyup', () => {
-      const temp = this.props.t('Editor.KeyUpLineNumber', {
-        lineNumber: parseInt(this._cm.getCursor().line + 1, 10)
+    cm.current.on('keyup', () => {
+      const temp = t('Editor.KeyUpLineNumber', {
+        lineNumber: parseInt(cm.current.getCursor().line + 1, 10)
       });
       document.getElementById('current-line').innerHTML = temp;
     });
 
-    this._cm.on('keydown', (_cm, e) => {
+    cm.current.on('keydown', (cmm, e) => {
       // Show hint
-      const mode = this._cm.getOption('mode');
+      const mode = cmm.getOption('mode');
       if (/^[a-z]$/i.test(e.key) && (mode === 'css' || mode === 'javascript')) {
-        this.showHint(_cm);
+        showHint(cmm);
       }
     });
-
-    this._cm.getWrapperElement().style[
-      'font-size'
-    ] = `${this.props.fontSize}px`;
-
-    this.props.provideController({
-      tidyCode: this.tidyCode,
-      showFind: this.showFind,
-      showReplace: this.showReplace,
-      getContent: this.getContent
-    });
-  }
-
-  componentWillUpdate(nextProps) {
-    // check if files have changed
-    if (this.props.files[0].id !== nextProps.files[0].id) {
-      // then need to make CodeMirror documents
-      this.initializeDocuments(nextProps.files);
-    }
-    if (this.props.files.length !== nextProps.files.length) {
-      this.initializeDocuments(nextProps.files);
+  }, [mergedState.autocompleteHinter]);
+
+  useEffect(() => {
+    if (file.id) {
+      prevProps.current = {
```

Comment:
looks like you're setting `prevProps.current` on every render down on line 556 so I'm not sure if you need this effect.

Reply:
See my latest i have changed lot of things.